examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,350 @@
+            glue_all += GLUE_METRICS[ckpt][task]
+            if task in large_tasks:
+                glue_large += GLUE_METRICS[ckpt][task]
+        # GLUE Large and GLUE All
+        output_line.append("{:.2f}".format(glue_large/len(large_tasks)))
+        output_line.append("{:.2f}".format(glue_all/len(tasks)))
+        tb.append(output_line)
+    
+    print(tabulate(tb, headers='firstrow'))
+
+''' Set up CUDA environment and process pool for given finetuning jobs. '''
+def setup_finetuning_jobs(finetune_tasks: List[str], save_ckpts: List[bool], ckpt_load_path: str, ckpt_filename: str,
+                                         save_folder: str, save_locally: bool, bucket: Optional[str] = None, parent_ckpt: Optional[str] = None, load_ignore_keys: Optional[List[str]] = None) -> None:
+    cuda_envs = init_cuda_queue(len(finetune_tasks))
+    num_tasks = len(finetune_tasks)
+    
+    ''' Callback function for metric collection. '''
+    def log_metrics(metric) -> None:
+        if parent_ckpt: 
+            logged_ckpt_name = parent_ckpt
+        else:
+            logged_ckpt_name = ckpt_filename
+
+        if logged_ckpt_name not in GLUE_METRICS.keys():
+                GLUE_METRICS[logged_ckpt_name] = {}
+        task = list(metric.keys())[0]
+        for m in metric[task]:
+            formatted_task = list(metric.keys())[0].split('_')[1] # remove "glue_" prefix
+            if formatted_task not in GLUE_METRICS[logged_ckpt_name].keys():
+                GLUE_METRICS[logged_ckpt_name][formatted_task] = metric[task][m].item()
+            else:
+                GLUE_METRICS[logged_ckpt_name][formatted_task] = sum([GLUE_METRICS[logged_ckpt_name][formatted_task], metric[task][m].item()]) / 2  # automatically average scores 
+        
+    if parent_ckpt: 
+        wandb_group_name = parent_ckpt
+    else:
+        wandb_group_name = ckpt_load_path
+
+    # finetuning from pretrained checkpoint
+    print(f"FINETUNING ON {ckpt_filename}!")
+    pool = mp.Pool(num_tasks, initializer=init_cuda_env, initargs=(cuda_envs,))
+    subprocs = []
+    for rank in range(num_tasks):
+        sp = pool.apply_async(train_finetune, (finetune_tasks[rank], save_ckpts[rank], ckpt_load_path, wandb_group_name, save_folder, save_locally, rank, bucket, load_ignore_keys), callback=log_metrics)
+        subprocs.append(sp)
+    for sp in subprocs:
+        sp.wait()
+    pool.close()
+    pool.join() # wait for join to proceed
+   
+    cuda_envs.close() 
+    cuda_envs.join_thread()
+    
+'''Run single instance of a finetuning job on given task. ''' 
+def train_finetune(task: str, save_ckpt: bool, load_path: str, wandb_group_name: str, save_folder: str, save_locally: bool, gpu_id: int, bucket: Optional[str] = None, load_ignore_keys: Optional[List[str]] = None):
+    # add dummy arg because arg parser removes first argument in tokenization
+    ft_hparams = TrainerHparams.create(cli_args=['-f',  f'./composer/yamls/models/glue/{task}.yaml',
```

Comment:
Sorry, what is "both" referring it?

Reply:
the train_pretrain and train_finetune

Reply:
Discussed offline -- difficult to do due since pretrain uses subprocess.run and finetune uses multiprocessing spawns. Code length and complexity will be cleaned up along with the other refactoring comments.